src/datasource.ts

```diff
@@ -17,4 +38,157 @@ export class DataSource extends DataSourceWithBackend<RedshiftQuery, RedshiftDat
+
+            if (isRedshiftCustomMeta(meta) && isRunning(meta.status)) {
+              queryID = meta.queryID;
+              status = meta.status;
+              this.storeQuery(target, { queryID });
+              return { ...target, queryID };
+            }
+          }
+
+          this.removeQuery(target);
+          return undefined;
+        },
+
+        /**
+         * The original request
+         */
+        query: (request: DataQueryRequest<RedshiftQuery>) => {
+          const { range, targets, requestId, intervalMs, maxDataPoints } = request;
+          const [_query] = targets;
+          const query: RedshiftQuery = {
+            ..._query,
+            ...(config.featureToggles.redshiftAsyncQueryDataSupport ? { meta: { queryFlow: 'async' } } : {}),
+          };
+
+          const data = {
+            queries: [
+              {
+                ...query,
+                intervalMs,
+                maxDataPoints,
+                datasource: this?.getRef(),
+                datasourceId: this.id,
+                ...this.applyTemplateVariables(query, request.scopedVars),
+              },
+            ],
+            range: range,
+            from: range.from.valueOf().toString(),
+            to: range.to.valueOf().toString(),
+          };
+
+          let headers = {};
+          if (isRunning(status)) {
+            headers = { 'X-Cache-Skip': true };
+          }
+          const options = {
+            method: 'POST',
+            url: '/api/ds/query',
+            data,
+            requestId,
+            headers,
+          };
+
+          return getBackendSrv()
+            .fetch<BackendDataSourceResponse>(options)
+            .pipe(map((result) => result.data))
+            .pipe(
+              map((r) => {
+                const frames: DataFrame[] = toDataQueryResponse({ data: r }).data;
+                return { data: frames };
+              })
+            );
+        },
+
+        /**
+         * Process the results
+         */
+        process: (data: DataFrame[]) => {
+          for (const frame of data) {
+            if (frame.fields.length > 0) {
+              allData.push(frame);
+            }
+          }
+
+          return allData;
+        },
+
+        shouldCancel: () => {
+          const { shouldCancel } = this.getQuery(target);
+          return !!shouldCancel;
+        },
+
+        /**
+         * Callback that gets executed when unsubscribed
+         */
+        onCancel: () => {
+          if (queryID) {
+            this.removeQuery(target);
+            this.postResource('cancel', {
+              queryId: queryID,
+            }).catch((err) => {
+              err.isHandled = true; // avoid the popup
+              console.log(`error cancelling query ID: ${queryID}`, err);
+            });
+          }
+        },
+      }
+    );
+  }
+
+  cancel(target: RedshiftQuery) {
+    this.storeQuery(target, { shouldCancel: true });
+  }
```

Comment:
This method is used by the query editor to cancel a query. It doesn't directly cancel the query. Instead, we store our intention to cancel the query in `this.runningQueries` by setting `shouldCancel` to true. This is because the code that is actually running the query is in the `requestLooper`. We're also storing our intention to cancel a query because we might not have the query id to cancel the query. This can happen if the user clicks the "run" button and then clicks "stop" right after and we haven't received a query id yet. That's why we let the `requestLooper` handle the cancellation because that's where we know when we have a query id.

Reply:
(also should probably be in the codebase)